<launch>
    <arg name="fcu_conn" default="usb"/>
    <arg name="fcu_ip" default="127.0.0.1"/>
    <arg name="gcs_bridge" default="tcp"/>
    <arg name="web_video_server" default="true"/>
    <arg name="rosbridge" default="true"/>
    <arg name="main_camera" default="true"/>
    <arg name="optical_flow" default="false"/>
    <arg name="aruco" default="false"/>
    <arg name="rc" default="true"/>
    <arg name="arduino" default="false"/>
    <arg name="vl53l1x" default="false"/>

    <!-- mavros -->
    <include file="$(find clever)/launch/mavros.launch">
        <arg name="fcu_conn" value="$(arg fcu_conn)"/>
        <arg name="fcu_ip" value="$(arg fcu_ip)"/>
        <arg name="gcs_bridge" value="$(arg gcs_bridge)"/>
        <arg name="viz" value="true"/>
    </include>

    <!-- web video server -->
    <node name="web_video_server" pkg="web_video_server" type="web_video_server" if="$(arg web_video_server)" required="false" respawn="true" respawn_delay="5"/>

    <!-- aruco vpe -->
    <include file="$(find clever)/launch/aruco.launch" if="$(arg aruco)"/>

    <!-- optical flow -->
    <node pkg="nodelet" type="nodelet" name="optical_flow" args="load clever/optical_flow nodelet_manager" if="$(arg optical_flow)" clear_params="true">
        <remap from="image" to="main_camera/image_raw"/>
        <remap from="camera_info" to="main_camera/camera_info"/>
    </node>

    <!-- main nodelet manager -->
    <node pkg="nodelet" type="nodelet" name="nodelet_manager" args="manager" output="screen" clear_params="true">
        <param name="num_worker_threads" value="2"/>
    </node>

    <node pkg="tf2_ros" type="static_transform_publisher" name="local_origin_upside_down_frame" args="0 0 0 3.1415926 3.1415926 0 local_origin local_origin_upside_down"/>


    <!-- simplified offboard control -->
    <node name="simple_offboard" pkg="clever" type="simple_offboard.py" output="screen"/>
    <!-- Auxiliary frames -->
    <node name="frames" pkg="clever" type="frames" output="screen">
        <param name="body/frame_id" value="fcu_horiz"/>
    </node>


    <!-- main camera -->
    <include file="$(find clever)/launch/main_camera.launch" if="$(arg main_camera)"/>

    <!-- rosbridge -->
    <include file="$(find rosbridge_server)/launch/rosbridge_websocket.launch" if="$(eval rosbridge or rc)"/>

    <!-- vl53l1x ToF rangefinder -->
    <node name="vl53l1x" pkg="clever" type="vl53l1x.py" output="screen" if="$(arg vl53l1x)">
        <param name="z_shift" value="-0.05"/>
        <!-- <remap from="~range" to="mavros/distance_sensor/rangefinder_3_sub"/> -->
    </node>

    <!-- rc backend -->
    <node name="rc" pkg="clever" type="rc" output="screen" if="$(arg rc)"/>

    <!-- Arduino bridge -->
    <include file="$(find clever)/launch/arduino.launch" if="$(arg arduino)"/>
</launch>
